File: src/test/java/pl/training/workshop/m6/s10_implicittree/S10EquivalenceTest.java
package pl.training.workshop.m6.s10_implicittree;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Stream;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import pl.training.workshop.support.Scene;

/** Niezależne oczekiwania (policzone ręcznie) dla ceny i wydruku zestawów. */
final class S10EquivalenceTest {
    @TestFactory
    Stream<DynamicTest> everyStepPricesAndRendersCombosTheSame() {
        return Scene.<List<?>, String>variants()
                .variant("start", safe(new pl.training.workshop.m6.s10_implicittree.start.BarMenu()::price,
                        new pl.training.workshop.m6.s10_implicittree.start.BarMenu()::render))
                .variant("step1", safe(new pl.training.workshop.m6.s10_implicittree.step1.BarMenu()::price,
                        new pl.training.workshop.m6.s10_implicittree.step1.BarMenu()::render))
                .variant("step2", safe(new pl.training.workshop.m6.s10_implicittree.step2.BarMenu()::price,
                        new pl.training.workshop.m6.s10_implicittree.step2.BarMenu()::render))
                .variant("step3", safe(new pl.training.workshop.m6.s10_implicittree.step3.BarMenu()::price,
                        new pl.training.workshop.m6.s10_implicittree.step3.BarMenu()::render))
                .expect("zestaw z podzestawem", List.of("Zestaw Duo", "Popcorn L=18.00",
                        List.of("Napoje", "Cola 0.5=9.00", "Cola 0.5=9.00"), "Nachos=14.00"), """
                        50.00
                        Zestaw Duo 50.00
                          Popcorn L 18.00
                          Napoje 18.00
                            Cola 0.5 9.00
                            Cola 0.5 9.00
                          Nachos 14.00
                        """)
                .expect("pusty zestaw", List.of("Pusty"), "0.00\nPusty 0.00\n")
                .expect("brak nazwy", List.of(), "ERROR combo needs a name")
                .expect("element nieobsługiwany", List.of("Zestaw", "Cola=9.00", 5), "ERROR unsupported element: 5")
                .expect("produkt bez ceny", List.of("Zestaw", "Cola"), "ERROR product needs a price: Cola")
                .tests();
    }

    static Function<List<?>, String> safe(Function<List<?>, ?> price, Function<List<?>, String> render) {
        return definition -> {
            try {
                return price.apply(definition) + "\n" + render.apply(definition);
            } catch (IllegalArgumentException exception) {
                return "ERROR " + exception.getMessage();
            }
        };
    }
}

File: csharp/tests/Training.Workshop.Tests/M6/S10ImplicitTree/S10EquivalenceTest.cs
using Training.Workshop.Tests.Support;

namespace Training.Workshop.Tests.M6.S10ImplicitTree;

/// <summary>Niezależne oczekiwania (policzone ręcznie) dla ceny i wydruku zestawów.</summary>
public sealed class S10EquivalenceTest
{
    private static readonly Scene<IReadOnlyList<object>, string> Scene = Support.Scene.Variants<IReadOnlyList<object>, string>()
        .Variant("start", Safe(new Training.Workshop.M6.S10ImplicitTree.Start.BarMenu().Price,
            new Training.Workshop.M6.S10ImplicitTree.Start.BarMenu().Render))
        .Variant("step1", Safe(new Training.Workshop.M6.S10ImplicitTree.Step1.BarMenu().Price,
            new Training.Workshop.M6.S10ImplicitTree.Step1.BarMenu().Render))
        .Variant("step2", Safe(new Training.Workshop.M6.S10ImplicitTree.Step2.BarMenu().Price,
            new Training.Workshop.M6.S10ImplicitTree.Step2.BarMenu().Render))
        .Variant("step3", Safe(new Training.Workshop.M6.S10ImplicitTree.Step3.BarMenu().Price,
            new Training.Workshop.M6.S10ImplicitTree.Step3.BarMenu().Render))
        .Expect("zestaw z podzestawem", ["Zestaw Duo", "Popcorn L=18.00",
            new object[] { "Napoje", "Cola 0.5=9.00", "Cola 0.5=9.00" }, "Nachos=14.00"], """
            50.00
            Zestaw Duo 50.00
              Popcorn L 18.00
              Napoje 18.00
                Cola 0.5 9.00
                Cola 0.5 9.00
              Nachos 14.00

            """)
        .Expect("pusty zestaw", ["Pusty"], "0.00\nPusty 0.00\n")
        .Expect("brak nazwy", [], "ERROR combo needs a name")
        .Expect("element nieobsługiwany", ["Zestaw", "Cola=9.00", 5], "ERROR unsupported element: 5")
        .Expect("produkt bez ceny", ["Zestaw", "Cola"], "ERROR product needs a price: Cola");

    public static TheoryData<string> Cases => Scene.Tests();

    [Theory]
    [MemberData(nameof(Cases))]
    public void EveryStepPricesAndRendersCombosTheSame(string test) => Scene.Run(test);

    internal static Func<IReadOnlyList<object>, string> Safe(
        Func<IReadOnlyList<object>, object> price, Func<IReadOnlyList<object>, string> render)
    {
        return definition =>
        {
            try
            {
                return price(definition) + "\n" + render(definition);
            }
            catch (ArgumentException exception)
            {
                return "ERROR " + exception.Message;
            }
        };
    }
}
